Replace Conv1d with Conv2d

taivu1998 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43221 Add benchmark for performance comparison
* **#42867 Replace Conv1d with Conv2d**

Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

Commit:
Replace Conv1d with Conv2d

[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"




[ghstack-poisoned]

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
+            else:
+                FileCheck().check_count(pattern, v, exactly=True).run(deserialized_scripted_model.graph)
+        xnnpack_result = deserialized_scripted_model(input_data)
+        torch.testing.assert_allclose(ref_result, xnnpack_result, rtol=1e-2, atol=1e-3)
```
Also run the model through optimize_for_mobile pass and check if tests still pass. This is mainly to transform the `conv2d` to prepacked counterparts of it that use xnnpack.

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
+                self.groups = groups
+
+            def forward(self, x):
+                return F.conv1d(x, self.weight, self.bias,
```
I would say let's do conv1d, relu, and then FC.

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
+        xnnpack_result = deserialized_scripted_model(input_data)
+        torch.testing.assert_allclose(ref_result, xnnpack_result, rtol=1e-2, atol=1e-3)
+
+    @given(batch_size=st.integers(0, 3),
+           input_channels_per_group=st.integers(1, 32),
+           height=st.integers(5, 64),
+           output_channels_per_group=st.integers(1, 32),
+           groups=st.integers(1, 16),
+           kernel=st.integers(1, 5),
+           strides=st.integers(1, 2),
+           paddings=st.integers(0, 2),
+           dilations=st.integers(1, 2))
```
I think test_xnnpack_integration.py needs to be fixed to not use hypothesis, because it generates only a handful of tests and no iterate through all comobinations.
If possible can we use https://docs.python.org/2/library/itertools.html#itertools.product, to get like all combinations? At least for some options like groups, kernel, strides, padding and dilations.

kimishpatel commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -24,6 +24,30 @@ namespace jit {
+
+  std::string conv_2d_pattern = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %padding_w : int = prim::ListUnpack(%padding)
```
Why only padding is adjusted like this? What about stride, and dilations?

taivu1998 replied:
I think we can handle stride and dilation similarly as well. However, it might not be needed since we only move in one dimension in conv1d, so the added dimension for stride and dilation may not matter too much. Padding is different, because we don't want to pad the added dimension.

kimishpatel replied:
I think we should be consistent and clear here. It is ok that conv2d might handle that gracefully, but it may change and not handle undefined value gracefully. We should avoid baking assumption about what some function might do, unless it is strongly coupled with our implementation.

kimishpatel commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -323,6 +363,16 @@ script::Module optimizeForMobile(
 
 #else
 
+void transformConv1dToConv2d(std::shared_ptr<Graph>& graph) {
+  TORCH_INTERNAL_ASSERT(
+      "XNNPACK is not enabled. Please build with USE_XNNPACK=1");
+}
+
+void transformConv1dToConv2d(script::Module& module) {
+  TORCH_INTERNAL_ASSERT(
+      "XNNPACK is not enabled. Please build with USE_XNNPACK=1");
+}
+
```
Actually, this is just rewriting graph into another graph that does not need xnnpack support to be executed. So we can remove these and move the other functions you added outside of ifdef.
Eventually we can think of moving them to a separate file once we have these working.

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
+                     " Please build with USE_XNNPACK=1.")
+class TestXNNPACKConv1dTransformPass(TestCase):
+    @staticmethod
+    def validate_transform_conv1d_to_conv2d(
```
I added this function in class because there was a reuse there. Since I dont see much reuse, you can inline this.

taivu1998 replied:
I thought we may need this function if we want to add more test cases later right?

kimishpatel replied:
Usually, we should introduce something only when we need it. But we can keep this for now. I am ok with it. You can add another test case that I described where we do conv1d+relu+FC etc.

kimishpatel commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -246,6 +270,22 @@ void runCanonicalOptimizations(script::Module& module) {
+
+void transformConv1dToConv2d(script::Module& module) {
+  for (auto& method : module.get_methods()) {
+    auto graph = method.graph();
```
`auto&` to avoid refcount bump.

taivu1998 replied:
I did try it, but it raised an error, so I just went for `auto`.

kimishpatel reviewed (changes requested):
Overall this looks good. Address some of the comments and think about adding benchmark. Good work!

Commit:
Update on "Replace Conv1d with Conv2d"




[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"




[ghstack-poisoned]

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
+            pattern_count_transformed_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": 1}
+            pattern_count_optimized_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": -1}
```
Also add check for prepacked::conv2d_clamp_run or something like that. See other tests for the pattern.

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
+                                 self.stride, self.padding, self.dilation, self.groups)
+                    x = F.relu(x)
+                    x = x.view(x.size(0), -1)
+                    x = F.linear(x, self.fc_weight, self.fc_bias)
```
Great. Thanks for adding this test.

kimishpatel commented on test/test_xnnpack_integration.py:
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
+            pattern_count_transformed_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": 1}
+            pattern_count_optimized_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": -1}
```
Similar to other comment. Add check for prepack stuff.

kimishpatel reviewed (changes requested):
This looks pretty good. Just one small change about adding additional check.

Commit:
Update on "Replace Conv1d with Conv2d"




[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"


Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"


Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"


Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"


Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

[ghstack-poisoned]

Commit:
Update on "Replace Conv1d with Conv2d"


Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

[ghstack-poisoned]
